src/unitxt/inference.py

```diff
@@ -35,9 +41,16 @@ def infer_log_probs(self, dataset):
         [ "top_tokens": [ "text": ..., "logprob": ...} , ... ]
 
         """
-        [self.verify_instance(instance) for instance in dataset]
+        for instance in dataset:
+            self.verify_instance(instance)
         return self._infer_log_probs(dataset)
 
+    def process(self, page: List[Dict], stream_name: Optional[str] = None) -> Generator:
```

Comment:
Right now, the process calls the regular inference call and not the log_prob inference.  
Perhaps it would be clearer and more flexible if we used composition , so we would have two classes:

InferenceEngineOperator(engine: InferenceEngine )
LogProbInferenceEngineOperator(engine: InferenceEngine )

which would implement the process() method ?

Reply:
Makes sense, I this more too, please mention it in the [unmerged PR](https://github.com/IBM/unitxt/pull/909).

Reply:
I actually think there are two options.  

1) Create two types of inference engine:

InferenceEngine
LogProbInferenceEngine

Then each can have a single infer method.  They can also be derived from an Operator with  a  "process" statement.

2) Keep  one type of inference engine and have two separator operators (as we discussed before)
 
Somehow, I now tend to the first option, but I'm not sure.
 
[BTW, there is also a request for adding EmbeddingEngine, which embeds each of the input ] 
 
 What do you think ? @jezekra1 , @OfirArviv , @elronbandel , @lilacheden ?

Reply:
> Somehow, I now tend to the first option, but I'm not sure.

I like it too because it's obvious which model supports which method instead of raising `NotImplementedError`

Reply:
I talked with Ofir  -  other libraries have one class for inference engine.  [e.g LM-eval harness  https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/api/model.py ]   It will also not require a separate class for input parameters when used in the different modes.

Reply:
After much deliberations, we think that there should be a separate Operator class.

class Infer(PagedStreamOperator)):  
   inference_engine: InferenceEngine
   to_field : str ="prediction"
   to_metadata_field : str = "generate_info"
   
   def process(self, page: List[Dict], stream_name: Optional[str] = None) -> Generator:
          for instance, output, metadata in zip(page, *self._infer(page)):
             instance[self.to_field] = output
             instance[self.to_metadata_field] = metadata
             yield instance